src/webhook_launcher/app/payload.py

```diff
@@ -429,3 +442,113 @@ def handle(self):
                         mapobj, payload["user"], lsr=seenrev, tag=tag
                     )
+
+
+class BbPushV2(Payload):
+    def __init__(self, data):
+        super(BbPushV2, self).__init__(data)
+        try:
+            data['push']
+            path = data['repository']['full_name']
+        except KeyError as e:
+            raise PayloadParsingError("Not a BbPushV2 payload: %s" % e)
+
+        print("bitbucket V2 payload")
+        self.url = urlparse.urljoin("https://bitbucket.org", path) + '.git'
+
+    def handle(self):
+        branches = {}
+        tags = {}
+        user = self.data['actor']['username']
+
+        # Collect branches and tags
+        for change in self.data['push']['changes']:
+            newref = change.get('new')
+            commits = change.get('commits', [])
+            if not newref:
+                # TODO: Handle delete events
+                continue
+            name = newref['name']
+            hash = newref['target']['hash']
+            if newref['type'] == 'branch':
+                branches[name] = [hash, commits]
+            elif newref['type'] == 'tag':
+                tags[name] = [hash, set()]
+            else:
+                print("Unknown ref type '%s'" % newref['type'])
+                continue
+
+        # Find branches for tags
+        for tag, (hash, tag_branches) in tags.items():
+            # TODO: this will only find branches with head matching the tag
+            bbcall = bbAPIcall(self.data['repository']['full_name'])
+            bts = bbcall.branches_tags()
+            for branch in bts['branches']:
+                if branch['changeset'] == hash:
+                    tags[tag][1].add(branch['name'])
+
+        # Handle commits in branches
+        for branch, (revision, commits) in branches.iteritems():
+            mapobjs = WebHookMapping.objects.filter(
+                repourl=self.url, branch=branch,
+            )
+            if not len(mapobjs):
+                packages = self.params.get("packages", None)
+                mapobjs = self.create_placeholder(
+                    self.url, branch, packages=packages
+                )
+            notified = False
+            for mapobj in mapobjs:
+                seenrev, created = LastSeenRevision.objects.get_or_create(
+                    mapping=mapobj
+                )
+
+                emails = set()
+                for commit in commits:
+                    emails.add(commit['author']['raw'])
+                    if len(emails) == 2:
+                        break
+                if emails:
+                    seenrev.emails = json.dumps(list(emails))
+
+                if created or seenrev.revision != revision:
+                    print(
+                        "%s in %s was not seen before, notify it if enabled" %
+                        (revision, branch)
+                    )
+                    seenrev.revision = revision
+                    seenrev.payload = json.dumps(self.data)
```

Comment:
This deserializing/serializing of the payload bothers me, as the only form of validation it is a bit thin, and otherwise why not just keep the original around?